### ISSUE TITLE
torch.func.grad hangs on nested differentiation

### ISSUE THREAD

**Issue by rtqichen** [pytorch/pytorch#121711]
### 🐛 Describe the bug

I want to compute nested derivatives of the form `d/ddx_1 d/dx_2 d/dx_3 ... f(x_1, x_2, x_3,...)`. This works with autograd.grad but torch.func.grad seems to hang.

MWE:

```
import torch
print(torch.__version__)
from torch.func import grad, jvp

d = 10
f = torch.prod
x = torch.randn(d).requires_grad_(True)

# Convert to indexable inputs.
def indexable_f(*args):
    return f(torch.stack(args))

# This allows us to take derivatives w.r.t. single elements.
inputs = list(x)
nested_deriv = indexable_f(*inputs)
for i in range(d):
    nested_deriv = torch.autograd.grad(nested_deriv, inputs[i], create_graph=True)[0]
print(nested_deriv)  # This works fine.

# This torch.func.grad version doesn't seem to work and runs forever.
nested_deriv = indexable_f
for i in range(d):
    nested_deriv = grad(nested_deriv, argnums=i)
nested_deriv(*inputs)  # THIS HANGS. 
```

### Versions

This is tested in colab notebook (https://colab.research.google.com/drive/1tbK-YYFVWMx6t6AulElap-JhInLpt-hT?usp=sharing) with the pytorch version 2.1.0+cu121.

**Comment by zou3519** [comment]
Hi-pri: Guilherme and I dug into this it looks like the the autograd graph is messed up in a bad way

**Comment by guilhermeleobas** [comment]
I have a small update on this. 

It turns out that PyTorch pushes an incredible large number of nodes to the autograd graph. It is actually computing the gradient, but the graph is so big that it doesn't finish in a reasonable amount of time:

```
[iter=1] nodes in graph: 2
[iter=2] nodes in graph: 14
[iter=3] nodes in graph: 205
[iter=4] nodes in graph: 1510
[iter=5] nodes in graph: 8906
[iter=6] nodes in graph: 44344
[iter=7] nodes in graph: 194988
[iter=8] nodes in graph: 776604
[iter=9] nodes in graph: 2849996
[iter=10] nodes in graph: ?? (my computer gave up before this gets computed)
```

**Comment by rtqichen** [comment]
Yeah, I've found that nested differentiation scales exponentially in compute time. (This would be a very nice test case for torch.compile as well.) But regardless, the torch.autograd.grad version seems to be able to compute it in reasonable time at least.

**Comment by guilhermeleobas** [comment]
`torch.autograd.grad` pushes fewer nodes to the graph:

```
[iter=1] nodes in graph: 3
[iter=2] nodes in graph: 6
[iter=3] nodes in graph: 12
[iter=4] nodes in graph: 39
[iter=5] nodes in graph: 148
[iter=6] nodes in graph: 575
[iter=7] nodes in graph: 2246
[iter=8] nodes in graph: 8821
[iter=9] nodes in graph: 34836
[iter=10] nodes in graph: 138195
```

**Comment by zou3519** [comment]
This is expected behavior. In the provided MWE, we are not comparing apples-to-apples. The following two things are comparable (and both hang):

```py
import torch
print(torch.__version__)
from torch.func import grad, jvp

d = 10
f = torch.prod
x = torch.randn(d).requires_grad_(True)

# Convert to indexable inputs.
def indexable_f(*args):
    return f(torch.stack(args))

# This allows us to take derivatives w.r.t. single elements.
inputs = list(x)
inputs = [t.detach().requires_grad_() for t in inputs]
nested_deriv = indexable_f(*inputs)
for i in range(d):
    nested_deriv = torch.autograd.grad(nested_deriv, inputs[i], create_graph=True)[0]
print(nested_deriv)  # hangs
```

```py
# This torch.func.grad version doesn't seem to work and runs forever.
nested_deriv = indexable_f
for i in range(d):
    nested_deriv = grad(nested_deriv, argnums=i)
nested_deriv(*inputs)  # THIS HANGS. 
```

Underneath the hood, torch.func creates new graph leaves. The equivalent thing using autograd.grad is `inputs = [t.detach().requires_grad_() for t in inputs]`.

**Comment by zou3519** [comment]
That being said, I wonder if there's anything we can make the functorch situation faster by rewriting the computation. Using JVP does result in something that doesn't hang:
```py
import torch
print(torch.__version__)
from torch.func import grad, jvp

d = 10
f = torch.prod
x = torch.randn(d).requires_grad_(True)

# Convert to indexable inputs.
def indexable_f(*args):
    return f(torch.stack(args))

# This allows us to take derivatives w.r.t. single elements.
inputs = list(x)
nested_deriv = indexable_f(*inputs)
for i in range(d):
    nested_deriv = torch.autograd.grad(nested_deriv, inputs[i], create_graph=True)[0]
print(nested_deriv)  # This works fine.

# JVP variant of this
def d_dxi(f, i):
    mask = torch.eye(d)[i]
    def get_grad(x):
        result, tangent = jvp(f, (x,), (mask,))
        return tangent
    return get_grad

nested_deriv = f
for i in range(d):
    nested_deriv = d_dxi(nested_deriv, i)
print(nested_deriv(x))
```

**Comment by guilhermeleobas** [comment]
I'll investigate further on why JVP doesn't hang.
